voc/python/ast.py

```diff
@@ -1712,7 +1712,8 @@ def visit_GeneratorExp(self, node):
 
     @node_visitor
     def visit_Yield(self, node):
-        self.visit(node.value)
+        if hasattr(node, "lineno"):
+            self.visit(node.value)
```

Comment:
It isn't clear to me why `lineno` is a relevant attribute here. Is there a type of AST node that *doesn't* have a linen that might be visited? 

At the very least, a comment explaining the need for this branch is required; better still would be a test case that exercises this case (assuming there isn't one already)

Reply:
This is sort of a 'hackish' way to differentiate a programmatically defined yield node from those generated normally by ast.parse. 

Combined with #823, there will be 3 types of yield node passed into visit_Yield method:
1.  Ordinary yield node generated from a line of code (eg `yield x`). In this case we need to visit x (the default implementation).
2. Ordinary yield node generated from a line of code, but without value/operand (eg `yield` or `x=yield`). In this case we push a `org.python.types.NoneType.NONE` object on stack without visiting the none value.
3. Programmatically defined yield node (in line 1778 `self.visit(ast.Yield(None))`) This node is created in visit_YieldFrom and the absent of `lineno` is use to indicate that a value has been pushed on stack hence there is no need to call self.visit(node.value)

Edit: I just remembered it is possible to add custom attribute to a class so I'm thinking of using a flag (i.e `skip_visit`) as custom attribute of the programmatically defined yield node instead of the `lineno` to increase clarity, what do you think?

Reply:
Ok - that reasoning makes sense; however, it isn't immediately obvious from the code. Adding a comment explaining what is going on would be helpful. A custom attribute isn't necessary; as long as the code explains what it's doing, that will be sufficient.